### PR TITLE
Fix `validate_aligned` edge-cases

### PR DESCRIPTION
**Pull request by giladchase**
- Handle empty Database edge cases on both sides.
- Rename `Unaligned` into `Misaligned` (a legal alignment always exists).
- Add annotations for readability

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 2 unresolved discussions (waiting on @alonh5 and @giladchase)

___
*[`crates/native_blockifier/src/errors.rs` line 67 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6YIntDBNTp82GQ3gF:-NR6YIntDBNTp82GQ3gG:b-u77ijn) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/errors.rs#L67)):*
> ```XML
>         "Blockifier storage is not aligned with the main storage; latest block ID in blockifier: \
>          {blockifier_latest_block_id}, latest block ID in main storage: {actual_latest_block_id}."
>     )]
> ```

Change `BigInt`s to `u64`.
___
*[`crates/native_blockifier/src/storage.rs` line 115 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6XytP5B_505X0CuPp:-NR6XytP5B_505X0CuPq:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L115)):*
> ```XML
> 
>     #[args(latest_block_id)]
>     pub fn validate_aligned(&self, latest_block_id: BigInt) -> NativeBlockifierResult<()> {
> ```





_Suggestion:_
```XML
latest_block_id: u64
```

**Review by alonh5 — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 4 unresolved discussions (waiting on @giladchase)

___
*[`crates/native_blockifier/src/storage.rs` line 129 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NRS-CxeCE15gjFMLdMo:-NRS-CxeCE15gjFMLdMp:b-tqq72x) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L129)):*
> ```XML
>         let current_block_number: u64 = next_block_number - 1;
>         let block_id = self
>             .get_block_id(current_block_number)?
> ```

Is there a reason we're returning the block ID as bytes and not a `StarkFelt` as it is?
___
*[`crates/native_blockifier/src/storage.rs` line 130 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NRRyUuj6fg2j9-zotVM:-NRRyUuj6fg2j9-zotVN:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L130)):*
> ```XML
>         let block_id = self
>             .get_block_id(current_block_number)?
>             .expect("blocks with a number should have an id");
> ```





_Suggestion:_
```XML
"Current block must have an ID."
```

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 4 unresolved discussions (waiting on @elintul)

___
*[`crates/native_blockifier/src/errors.rs` line 67 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6YIntDBNTp82GQ3gF:-NRXOV2WB7IcGMGRtSZb:b-al4l1p) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/errors.rs#L67)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Change `BigInt`s to `u64`.
</blockquote></details>

Due to the genesis block edge-case, in which the block_id is `-1`, we have to keep `latest_block_id` as a signed number.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 4 unresolved discussions (waiting on @elintul)

___
*[`crates/native_blockifier/src/storage.rs` line 115 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6XytP5B_505X0CuPp:-NRXPlqG0yBkp4XzUebL:bc9au0k) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L115)):*
> ```XML
> 
>     #[args(latest_block_id)]
>     pub fn validate_aligned(&self, latest_block_id: BigInt) -> NativeBlockifierResult<()> {
> ```

 Ditto to the above.

**Review by elintul — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 4 unresolved discussions (waiting on @giladchase)

___
*[`crates/native_blockifier/src/errors.rs` line 67 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6YIntDBNTp82GQ3gF:-NRXQ9OTASjVfvuiJXTd:b-428jbx) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/errors.rs#L67)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Due to the genesis block edge-case, in which the block_id is `-1`, we have to keep `latest_block_id` as a signed number.
</blockquote></details>

But we end up passing 64bit all-ones.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 4 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/src/errors.rs` line 67 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6YIntDBNTp82GQ3gF:-NRYmYts8ywjZPPhnrVP:b-cvgily) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/errors.rs#L67)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

But we end up passing 64bit all-ones.
</blockquote></details>

Currently only previous_block_id uses that trick, whereas `validate_aligned` expects the -1 value to be passed as is.

We talked about passing `None` instead of -1, which will require a small refactor in Rust as well; I'm blocking this PR and downgrading it to draft until that's settled.
___
*[`crates/native_blockifier/src/storage.rs` line 129 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NRS-CxeCE15gjFMLdMo:-NRYpLt6BrUyOtYy5p_j:but1i4y) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L129)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Is there a reason we're returning the block ID as bytes and not a `StarkFelt` as it is?
</blockquote></details>

That's what papyrus return :(
`get_block_id` returns `Starkhash` type, which is a tuple struct around bytes.
We can cast it `into` a `FieldElement` ourselves, but that won't help us compare with a BigInt (we'll still need an additional cast either of it or of the BigInt), so might as well cast the `BigInt` into bytes. 
___
*[`crates/native_blockifier/src/storage.rs` line 130 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NRRyUuj6fg2j9-zotVM:-NRYq4_u19cyxaaQZ-1B:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L130)):*
> ```XML
>         let block_id = self
>             .get_block_id(current_block_number)?
>             .expect("blocks with a number should have an id");
> ```

Done.

**Review by elintul — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @giladchase)

___
*[`crates/native_blockifier/src/storage.rs` line 115 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6XytP5B_505X0CuPp:-NRbAbXrCeCLXeYfAXwK:b-qhesnz) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L115)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

 Ditto to the above.
</blockquote></details>

We wanted to make it `Option<u64>`, right?

**Review by elintul — COMMENTED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 2 unresolved discussions (waiting on @giladchase)

___
*[`crates/native_blockifier/src/storage.rs` line 115 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6XytP5B_505X0CuPp:-NRlasds5scMLkBdU8Yz:b9ah8tx) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L115)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

ya, but i'm keeping this as draft until that change get's merged (it's already in PR)
</blockquote></details>

I think it's merged, you can go ahead.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/353)* status: all files reviewed, 2 unresolved discussions (waiting on @elintul)

___
*[`crates/native_blockifier/src/storage.rs` line 115 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/353#-NR6XytP5B_505X0CuPp:-NRmMhLt58tiQeTcxVYe:biuwdtk) ([raw file](https://github.com/starkware-libs/blockifier/blob/20002638b485f5b50525a2526bc2c2b4e4533aa6/crates/native_blockifier/src/storage.rs#L115)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I think it's merged, you can go ahead.
</blockquote></details>

I think https://reviewable.io/reviews/starkware-libs/blockifier/378#- makes this PR no longer relevant...
Once merged I'll close this.

**Pull request closed without merging**
